src/api/controllers/cache-controller.ts

```diff
@@ -0,0 +1,193 @@
+import { RequestHandler, Request, Response } from 'express';
+import * as prom from 'prom-client';
+import { FoundOrNot, logger } from '../../helpers';
+import { DataStore, DbChainTip } from '../../datastore/common';
+import { asyncHandler } from '../async-handler';
+
+const CACHE_OK = Symbol('cache_ok');
+const CHAIN_TIP_LOCAL = 'chain_tip';
+
+interface ChainTipCacheMetrics {
+  chainTipCacheHits: prom.Counter<string>;
+  chainTipCacheMisses: prom.Counter<string>;
+  chainTipCacheNoHeader: prom.Counter<string>;
+}
+
+let _chainTipMetrics: ChainTipCacheMetrics | undefined;
+function getChainTipMetrics(): ChainTipCacheMetrics {
+  if (_chainTipMetrics !== undefined) {
+    return _chainTipMetrics;
+  }
+  const metrics: ChainTipCacheMetrics = {
+    chainTipCacheHits: new prom.Counter({
+      name: 'chain_tip_cache_hits',
+      help: 'Total count of requests with an up-to-date chain tip cache header',
+    }),
+    chainTipCacheMisses: new prom.Counter({
+      name: 'chain_tip_cache_misses',
+      help: 'Total count of requests with a stale chain tip cache header',
+    }),
+    chainTipCacheNoHeader: new prom.Counter({
+      name: 'chain_tip_cache_no_header',
+      help: 'Total count of requests that did not provide a chain tip header',
+    }),
+  };
+  _chainTipMetrics = metrics;
+  return _chainTipMetrics;
+}
+
+export function setResponseNonCacheable(res: Response) {
+  res.removeHeader('Cache-Control');
+  res.removeHeader('ETag');
+}
+
+/**
+ * Sets the response `Cache-Control` and `ETag` headers using the chain tip previously added
+ * to the response locals.
+ * Uses the latest unanchored microblock hash if available, otherwise uses the anchor
+ * block index hash.
+ */
+export function setChainTipCacheHeaders(res: Response) {
+  const chainTip: FoundOrNot<DbChainTip> | undefined = res.locals[CHAIN_TIP_LOCAL];
+  if (!chainTip) {
+    logger.error(
```

Comment:
Can this be a warning instead of an error? I guess we'll hit this path very often when the cache tip is still being generated, right?

Reply:
This condition would only be hit from buggy cache-control code. There should be an object at `res.locals[CHAIN_TIP_LOCAL]` even if it's a `found: false` chain tip.